src/hotspot/share/gc/g1/heapRegion.cpp

```diff
@@ -459,34 +457,99 @@ void HeapRegion::print_on(outputStream* st) const {
   st->print_cr("");
 }
 
-class VerifyRemSetClosure : public G1VerificationClosure {
+static bool is_oop_safe(oop obj) {
+  if (!oopDesc::is_oop(obj)) {
+    log_error(gc, verify)(PTR_FORMAT " not an oop", p2i(obj));
+    return false;
+  }
+
+  // Now examine the Klass a little more closely.
+  Klass* klass = obj->klass_raw();
+
+  bool is_metaspace_object = Metaspace::contains(klass);
+  if (!is_metaspace_object) {
+    log_error(gc, verify)("klass " PTR_FORMAT " of object " PTR_FORMAT " "
+                          "not metadata", p2i(klass), p2i(obj));
+    return false;
+  } else if (!klass->is_klass()) {
+    log_error(gc, verify)("klass " PTR_FORMAT " of object " PTR_FORMAT " "
+                          "not a klass", p2i(klass), p2i(obj));
+    return false;
+  }
+
+  return true;
+}
+
+// Closure that glues together validity check for oop references (first),
+// then optionally verifies the remembered set for that reference.
+class G1VerifyLiveAndRemSetClosure : public BasicOopIterateClosure {
+  using CardValue = CardTable::CardValue;
+
+  G1CollectedHeap* _g1h;
+  VerifyOption _vo;
+  bool _verify_remsets;
+
+  oop _containing_obj;
+
+  size_t _num_failures;
+
   G1CardTable *_ct;
 
-  template <class T>
-  void do_oop_work(T* p) {
-    assert(_containing_obj != nullptr, "Precondition");
-    assert(!_g1h->is_obj_dead_cond(_containing_obj, _vo), "Precondition");
+  bool has_failures() const { return _num_failures != 0; }
 
-    T heap_oop = RawAccess<>::oop_load(p);
-    if (CompressedOops::is_null(heap_oop)) {
-      return;
-    }
+  void print_object(outputStream* out, oop obj) {
+#ifdef PRODUCT
+    obj->print_name_on(out);
+#else // PRODUCT
+    obj->print_on(out);
+#endif // PRODUCT
+  }
+
+  template<class T>
+  bool verify_liveness(T* p, oop obj) {
+    bool is_in_heap = _g1h->is_in(obj);
 
-    oop obj = CompressedOops::decode_raw_not_null(heap_oop);
+    if (!is_in_heap || _g1h->is_obj_dead_cond(obj, _vo)) {
+      ResourceMark rm;
+      Log(gc, verify) log;
+      LogStream ls(log.error());
 
+      MutexLocker x(G1RareEvent_lock, Mutex::_no_safepoint_check_flag);
```

Comment:
Having this lock avoids the log messages for a single failures (these are multiple lines each) to not mix. Otherwise it would be very hard to analyze the failures. Does that make sense?

Reply:
Oh, you are right. For some reason I thought LogStream did that, but
apparently I misremembered. Some documentation for LogStream &etc would be
helpful.

One can avoid the interleaving by using LogMessage and
NonInterleavingLogStream, though that might end up generating more allocations
than a LogStream with the surrounding lock.